conda_models/repodata_patch.py

```diff
@@ -1,3 +1,24 @@
 """
 WIP
 """
+from pydantic import Field
+
+from ._base import ExtrasForbiddenModel
+from .repodata import AllOptionalRepodataRecord
+from .types import CondaPackageFileNameStr, PackageFileNameStr, TarBz2PackageFileNameStr
+
+
+class RepodataPatchInstructions(ExtrasForbiddenModel):
+    packages: dict[TarBz2PackageFileNameStr, AllOptionalRepodataRecord]
+    "The .tar.bz2 packages in the repodata that will be patched"
+    packages_conda: dict[CondaPackageFileNameStr, AllOptionalRepodataRecord] = Field(
```

Comment:
Note that it doesnt make sense to patch all fields from a RepodataRecord. Changing the name of a package doesnt make sense and changing the hash might break cache implementations. In Rattler we explicity limit the fields that can be patched for that reason. We also didnt encounter fields not covered by our implementation. Might be worth considering!

Reply:
Agreed that it should be prevented and it's A Bad Idea (tm). 

- I don't see anywhere in the Anaconda code where this is properly enforced though. [In principle only these fields](https://github.com/AnacondaRecipes/repodata-hotfixes/blob/3c1a3f6d58096bdf8f17d657b6c6b143917165d7/main.py#L514-L522), but then there are some exceptions right below... e.g. timestamps, build_number... 😬)
- In conda-forge, [_anything_ goes, even new filenames?](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/blob/6cd9e90b43fac9a027a0f08de1c0cc47875cdabe/recipe/gen_patch_json.py#L319-L334)

So while the file format seems to accept everything, there are some guidelines place for "reasonable" use and it looks like this is mostly addressed in code reviews, but sometimes we have exceptions. 🤷 

I've been thinking whether we want to get descriptive or prescriptive in this PR, and I have mixed feelings about it. For scalars, I've been trying to be more prescriptive, but with structs... just reflecting what (I think) is in use these days. I am not saying I won't change my mind about it, just writing down what the approach have been so far :D

Thanks for the feedback!

Also, are we sure we have a `packages.conda` field? Maybe [the readme is outdated](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/tree/main/recipe), and the seed dict is too?

Reply:
> I've been thinking whether we want to get descriptive or prescriptive in this PR, and I have mixed feelings about it. For scalars, I've been trying to be more prescriptive, but with structs... just reflecting what (I think) is in use these days. I am not saying I won't change my mind about it, just writing down what the approach have been so far :D

I understand where you're coming from. With rattler we definitely choose the prescriptive approach because I feel like there is way too much "legacy" code and supporting all those exceptions is simply harder. But the same logic does not necessarily hold for this project.

The `packages.conda` field is definitely used by the `conda-forge-repodata-patches` package. The readme does indeed seem to be outdated. Here is a small excerpt from the latest `linux-64` patches:

```json
 "packages.conda": {
    "4ti2-1.6.9-hbc9de56_1.conda": {
      "license_family": "GPL"
    },
    "acctaudem-1.0.1-h69042ef_3.conda": {
      "license_family": "GPL"
    },
    "actions-runner-2.299.1-he0ac6c6_0.conda": {
      "license_family": "MIT"
    },
    "actions-runner-2.300.0-h0cdce71_0.conda": {
      "license_family": "MIT"
    },
```

Reply:
Looking at the current state, I thin we should be **descriptive** in the initial implementation: The first step in this effort is to capture what is used currently so that people and projects can refer to the schema to ensure that their software is implemented correctly.

In further steps, we could get more and more prescriptive.

This brings me to a second thing that we'll need to have: versioning. As with everything, at some point, the schema will change.

To enable implementations to reference "the schema" they're using, we'll need to version the schemas. If we don't, we effectively can't change them without breaking something (be that software or user trust).